### ISSUE TITLE
chromosome with integer and binary values

### ISSUE THREAD

**Issue by antoniogarmendia** [jenetics/jenetics#696]
Hi,

I have been searching for a while, but I did not find a good way to define a chromosome which contains integers (part one) and binary values (part two) (e.g. chromosome = 51324 0101). Specifically, the first part should be EnumGene<Integer>. What would be the best way to define this chromosome?

The only way that I can think of is a wrapper of both Genes (BitGene and IntegerGene) that extends a CustomGene, and after that, implement a CustomChromosome, but I am afraid that probably there is a better way of doing this.

Thanks in advance!

Kind regards,
Antonio

**Comment by jenetics** [comment]
Your request seems to be similar to #562 and #116. A full example is given [here](https://github.com/jenetics/jenetics/blob/master/jenetics.example/src/main/java/io/jenetics/example/MixedGenotype.java). The better solution would be to use a encoding which work with only one gene type. But without additional information I'm not able to given you some guidance.

Regards
Franz

**Comment by antoniogarmendia** [comment]
Hi,

It seems that I did not use the good words to find those issues. I just tested the full example and it seems perfect for me. Thanks! 

The thing is that I am trying to implement the encoding proposed in this paper for hierarchical clustering. [[Rizzi] Genetic operators for hierarchical graph clustering.pdf](https://github.com/jenetics/jenetics/files/4601380/Rizzi.Genetic.operators.for.hierarchical.graph.clustering.pdf) In terms of implementation I thought that it would be good to have P1 (Vertex) and P2 (Separators). For the first part, it could be applied PartiallyMatchedCrossover and SwapMutator and for Part2 SinglePointCrossOver and a flip mutator. What do you think? 

Kind regards,
Antonio

**Comment by antoniogarmendia** [comment]
Hi, 

Given the description of my previous comment, I just prepare a minimal example based on [MixedMOEAO](https://github.com/jenetics/jenetics/blob/master/jenetics.example/src/main/java/io/jenetics/example/MixedMOEAOptimization.java) and [MixedGenotype](https://github.com/jenetics/jenetics/blob/master/jenetics.example/src/main/java/io/jenetics/example/MixedGenotype.java). The problem that I am facing now is the implementation of the alterers. I don't know how to apply the alteration methods to P1 and P2 separately and then join both results that will conform to the new population. 

I am working in this minimal example. [moo-sample-mixedgenotype.zip](https://github.com/jenetics/jenetics/files/4605916/moo-sample-mixedgenotype.zip)
I really do not know how to implement the Alterer interface in a good way (MixedChromosomeAlterer) I also implement the Problem interface in this way `Problem<List<Chromosome<AnyGene<?>>>, AnyGene<?>, Vec<double[]>>` the use of List<Chromosome> is correct? What about `AnyGene<?>`?

Would you please take a look at the code? Thanks in advance.

Kind regards,
Antonio

**Comment by jenetics** [comment]
You may have a look at the [PartialAlterer](https://jenetics.io/javadoc/jenetics/io/jenetics/PartialAlterer.html) class. This allows you to specify an alterer to only one chromosome of your mixed genotype.

**Comment by antoniogarmendia** [comment]
Hi Franz,

Sorry for the late reply. I just made an example and it seems to work very well. Find it attached jut in case you have any comments. I am very grateful to you for all your help.

[SampleMixedGenotypePartialAlterers.zip](https://github.com/jenetics/jenetics/files/4651786/SampleMixedGenotypePartialAlterers.zip)

Thanks!

Kind regards,
Antonio

**Comment by jenetics** [comment]
Looks good, if you want take this route ;-) Some simplifications are still possible.
```java
@SuppressWarnings({ "unchecked", "rawtypes" })
public class SampleMixedGenotypePartialAlterers {

    private static final Genotype ENCODING = Genotype.of(
        (Chromosome)PermutationChromosome.of(
            IntStream.rangeClosed(1,5).boxed()
                .collect(ISeq.toISeq())
        ),
        (Chromosome)BitChromosome.of(4)
    );

    public static void main(String[] args) {
        final Engine engine = Engine
                .builder(SampleMixedGenotypePartialAlterers::fitness, ENCODING)
                .alterers(
                    PartialAlterer.of(new PartiallyMatchedCrossover(0.4), 0),
                    PartialAlterer.of(new MultiPointCrossover(0.4), 1),
                    new Mutator<>(0.1))
                .survivorsSelector(NSGA2Selector.ofVec())
                .populationSize(200)
                .build();

        final ISeq<Phenotype> best = (ISeq<Phenotype>)engine.stream()
                .limit(300)
                .collect(MOEA.toParetoSet(IntRange.of(30, 50)));

        System.out.println(best);
    }

    ...
}
```

**Comment by antoniogarmendia** [comment]
Hi,

Thanks! I will add these simplifications.

Kind regards,
Antonio

**Comment by antoniogarmendia** [comment]
Hi,

I started to implement this approach on my real problem and, I realise that the results of the fitness functions are not working as expected. Then, I come back to our minimal example and it seems to me that the result of the fitness functions saved in the Phenotype is not equal to its further execution.

![ff-example](https://user-images.githubusercontent.com/9051720/82725282-37bb9580-9cdc-11ea-9ec9-bb694dac4c6e.png)

Find attached the example. I just added a static method that returns the bitCount.
[SampleMixedGenotypePartialAlterers.zip](https://github.com/jenetics/jenetics/files/4671382/SampleMixedGenotypePartialAlterers.zip)

Would you give it a try? Thanks in advance.

Kind regards,
Antonio

**Comment by jenetics** [comment]
Really strange, but I can't reproduce it. Which version of the library are you using?

**Comment by antoniogarmendia** [comment]
Hi,

I am using 5.2.0. I think that I am going to update it now and test it again and let you know.

Thanks!

Kind regards,
Antonio

**Comment by jenetics** [comment]
Maybe this bug (#663) is responsible for this behavior. I fixed it in version 6.0.0.

**Comment by antoniogarmendia** [comment]
Hi,

Thanks! It seems so, because I am testing with the 6.0.0 and this behaviour disappear. 

Kind regards,
Antonio
